html/captive-portal/lib/captiveportal/PacketFence/DynamicRouting/Module/Authentication/Login.pm

```diff
@@ -30,8 +30,22 @@ has '+sources' => (isa => 'ArrayRef['.join('|', @{sources_classes()}).']');
 
 has '+multi_source_object_classes' => (default => sub{sources_classes()});
 
+has 'challenge_template' => (is => 'ro', default => sub { "challenge.html" } );
+
+has 'challenge_data' => (is => 'rw', builder => '_build_challenge_data', lazy => 1, trigger => \&_trigger_challenge_data);
+
+sub _build_challenge_data {
+    my ($self) = @_;
+    return $self->app->session->{challenge_data};
```

Comment:
So example:

I have a chained login:
- First is ZammitCorp RADIUS login - uses module zammitcorp_login - a RADIUS login with challenge
- Second is JamaisseInc LDAP login - uses module bug_killer_3000

If you put the challenge in the global session then when you login on ZammitCorp it will put the challenge data in the global session meaning the JamaisseInc module will try to use it and go through the challenge code path.

If you put it in the module session, then the challenge data of ZammitCorp stays encapsulated into the module and the JamaisseInc authentication won't see it and won't go through the challenge codepath.

So the session of the module was made for that reason, encapsulating the state of the module into itself so data doesn't clash between modules.